File: sdk/python/tirisfal.py
"""
Tirisfal Secrets Manager Python SDK

Simple client for accessing secrets from Tirisfal.

Usage:
    from tirisfal import TirisfalClient

    client = TirisfalClient(
        server="https://your-worker.workers.dev",
        token="your-machine-account-token"
    )

    # Get a secret
    secret = client.get_secret("API_KEY", project_id="xxx", environment="prod")
    print(secret.value)
"""

import json
from dataclasses import dataclass
from typing import Optional
from urllib.request import Request, urlopen
from urllib.error import HTTPError


@dataclass
class Secret:
    """Represents a secret value."""
    id: str
    name: str
    value: str
    project_id: str
    environment: str
    note: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


@dataclass
class Project:
    """Represents a project."""
    id: str
    name: str
    description: Optional[str] = None
    user_id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


@dataclass
class MachineAccount:
    """Represents a machine account."""
    id: str
    name: str
    user_id: Optional[str] = None
    status: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TirisfalError(Exception):
    """Base exception for Tirisfal SDK errors."""
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.status_code = status_code


class TirisfalClient:
    """
    Client for Tirisfal Secrets Manager.

    Args:
        server: Base URL of the Tirisfal server
        token: Machine account access token or user JWT token
    """

    def __init__(self, server: str, token: str):
        self.server = server.rstrip("/")
        self.token = token

    def _request(
        self,
        method: str,
        path: str,
        data: Optional[dict] = None,
        params: Optional[dict] = None
    ) -> dict:
        """Make an HTTP request to the API."""
        url = f"{self.server}{path}"

        if params:
            query_string = "&".join(f"{k}={v}" for k, v in params.items() if v)
            if query_string:
                url += f"?{query_string}"

        headers = {
            "Authorization": f"Bearer {self.token}",
            "Content-Type": "application/json",
        }

        body = json.dumps(data).encode() if data else None
        req = Request(url, data=body, headers=headers, method=method)

        try:
            with urlopen(req) as response:
                return json.loads(response.read().decode())
        except HTTPError as e:
            error_body = e.read().decode() if e.fp else str(e)
            try:
                error_data = json.loads(error_body)
                message = error_data.get("error", error_body)
            except json.JSONDecodeError:
                message = error_body
            raise TirisfalError(message, e.code)

    # ==================== Secrets ====================

    def get_secret(
        self,
        name: str,
        project_id: str,
        environment: str = "prod"
    ) -> Secret:
        """
        Get a secret by name.

        Args:
            name: Secret name
            project_id: Project ID
            environment: Environment (prod, staging, dev, test)

        Returns:
            Secret object with decrypted value
        """
        data = self._request(
            "GET",
            f"/api/secrets/by-name/{name}",
            params={"project_id": project_id, "environment": environment}
        )
        return Secret(
            id=data["id"],
            name=data["name"],
            value=data["value"],
            project_id=data["project_id"],
            environment=data["environment"],
            note=data.get("note"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
        )

    def get_secrets(
        self,
        project_id: str,
        environment: Optional[str] = None
    ) -> list[Secret]:
        """
        List secrets in a project.

        Args:
            project_id: Project ID
            environment: Filter by environment (optional)

        Returns:
            List of Secret objects
        """
        params = {"project_id": project_id}
        if environment:
            params["environment"] = environment

        data = self._request("GET", "/api/secrets", params=params)
        return [
            Secret(
                id=s["id"],
                name=s["name"],
                value=s.get("value", ""),
                project_id=s["project_id"],
                environment=s["environment"],
                note=s.get("note"),
                created_at=s.get("created_at"),
                updated_at=s.get("updated_at"),
            )
            for s in data.get("data", [])
        ]

    def create_secret(
        self,
        name: str,
        value: str,
        project_id: str,
        environment: str = "prod",
        note: Optional[str] = None
    ) -> Secret:
        """
        Create a new secret.

        Args:
            name: Secret name
            value: Secret value
            project_id: Project ID
            environment: Environment
            note: Optional note

        Returns:
            Created Secret object
        """
        data = self._request(
            "POST",
            "/api/secrets",
            data={
                "name": name,
                "value": value,
                "project_id": project_id,
                "environment": environment,
                "note": note,
            }
        )
        return Secret(
            id=data["id"],
            name=data["name"],
            value=value,
            project_id=data["project_id"],
            environment=data["environment"],
            note=data.get("note"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
        )

    def update_secret(
        self,
        secret_id: str,
        value: Optional[str] = None,
        note: Optional[str] = None
    ) -> Secret:
        """
        Update an existing secret.

        Args:
            secret_id: Secret ID
            value: New value (optional)
            note: New note (optional)

        Returns:
            Updated Secret object
        """
        update_data = {}
        if value is not None:
            update_data["value"] = value
        if note is not None:
            update_data["note"] = note

        data = self._request("PUT", f"/api/secrets/{secret_id}", data=update_data)
        return Secret(
            id=data["id"],
            name=data["name"],
            value=value or data.get("value", ""),
            project_id=data["project_id"],
            environment=data["environment"],
            note=data.get("note"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
        )

    def delete_secret(self, secret_id: str) -> bool:
        """
        Delete a secret.

        Args:
            secret_id: Secret ID

        Returns:
            True if deleted successfully
        """
        self._request("DELETE", f"/api/secrets/{secret_id}")
        return True

    # ==================== Projects ====================

    def get_projects(self) -> list[Project]:
        """
        List all projects.

        Returns:
            List of Project objects
        """
        data = self._request("GET", "/api/projects")
        return [
            Project(
                id=p["id"],
                name=p["name"],
                description=p.get("description"),
                user_id=p.get("user_id"),
                created_at=p.get("created_at"),
                updated_at=p.get("updated_at"),
            )
            for p in data.get("data", [])
        ]

    def get_project(self, project_id: str) -> Project:
        """
        Get a project by ID.

        Args:
            project_id: Project ID

        Returns:
            Project object
        """
        data = self._request("GET", f"/api/projects/{project_id}")
        return Project(
            id=data["id"],
            name=data["name"],
            description=data.get("description"),
            user_id=data.get("user_id"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
        )

    # ==================== Machine Accounts ====================

    def get_machine_accounts(self) -> list[MachineAccount]:
        """
        List all machine accounts.

        Returns:
            List of MachineAccount objects
        """
        data = self._request("GET", "/api/machine-accounts")
        return [
            MachineAccount(
                id=m["id"],
                name=m["name"],
                user_id=m.get("user_id"),
                status=m.get("status"),
                created_at=m.get("created_at"),
                updated_at=m.get("updated_at"),
            )
            for m in data.get("data", [])
        ]

    # ==================== Health Check ====================

    def health_check(self, detailed: bool = False) -> dict:
        """
        Check server health.

        Args:
            detailed: Include detailed service checks

        Returns:
            Health status dict
        """
        params = {"detailed": "true"} if detailed else None
        return self._request("GET", "/health", params=params)


# Convenience function
def get_secret(
    server: str,
    token: str,
    name: str,
    project_id: str,
    environment: str = "prod"
) -> str:
    """
    Quick function to get a secret value.

    Args:
        server: Tirisfal server URL
        token: Access token
        name: Secret name
        project_id: Project ID
        environment: Environment

    Returns:
        Secret value as string
    """
    client = TirisfalClient(server, token)
    secret = client.get_secret(name, project_id, environment)
    return secret.value

File: sdk/go/tirisfal.go
// Package tirisfal provides a client for the Tirisfal Secrets Manager API.
//
// Usage:
//
//	client := tirisfal.NewClient("https://your-worker.workers.dev", "your-token")
//	secret, err := client.GetSecret("API_KEY", "project-id", "prod")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	fmt.Println(secret.Value)
package tirisfal

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Secret represents a secret value.
type Secret struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Value       string  `json:"value"`
	ProjectID   string  `json:"project_id"`
	Environment string  `json:"environment"`
	Note        *string `json:"note,omitempty"`
	CreatedAt   *string `json:"created_at,omitempty"`
	UpdatedAt   *string `json:"updated_at,omitempty"`
}

// Project represents a project.
type Project struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
	UserID      *string `json:"user_id,omitempty"`
	CreatedAt   *string `json:"created_at,omitempty"`
	UpdatedAt   *string `json:"updated_at,omitempty"`
}

// MachineAccount represents a machine account.
type MachineAccount struct {
	ID        string  `json:"id"`
	Name      string  `json:"name"`
	UserID    *string `json:"user_id,omitempty"`
	Status    *string `json:"status,omitempty"`
	CreatedAt *string `json:"created_at,omitempty"`
	UpdatedAt *string `json:"updated_at,omitempty"`
}

// HealthStatus represents the health status of the server.
type HealthStatus struct {
	Status    string `json:"status"`
	Timestamp string `json:"timestamp"`
	Version   string `json:"version"`
	Services  struct {
		Database string `json:"database"`
		Storage  string `json:"storage"`
	} `json:"services"`
}

// ErrorResponse represents an error response from the API.
type ErrorResponse struct {
	Error            string `json:"error"`
	ErrorDescription string `json:"error_description"`
}

// Client is the Tirisfal API client.
type Client struct {
	Server string
	Token  string
	HTTP   *http.Client
}

// NewClient creates a new Tirisfal client.
func NewClient(server, token string) *Client {
	return &Client{
		Server: strings.TrimRight(server, "/"),
		Token:  token,
		HTTP: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// request makes an HTTP request to the API.
func (c *Client) request(method, path string, body interface{}, params map[string]string) ([]byte, error) {
	u := c.Server + path

	if len(params) > 0 {
		q := url.Values{}
		for k, v := range params {
			if v != "" {
				q.Set(k, v)
			}
		}
		if encoded := q.Encode(); encoded != "" {
			u += "?" + encoded
		}
	}

	var bodyReader io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal body: %w", err)
		}
		bodyReader = strings.NewReader(string(jsonBody))
	}

	req, err := http.NewRequest(method, u, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+c.Token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		var errResp ErrorResponse
		if json.Unmarshal(respBody, &errResp) == nil && errResp.Error != "" {
			return nil, &APIError{
				StatusCode: resp.StatusCode,
				Message:    errResp.Error,
			}
		}
		return nil, &APIError{
			StatusCode: resp.StatusCode,
			Message:    string(respBody),
		}
	}

	return respBody, nil
}

// APIError represents an API error.
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("API error %d: %s", e.StatusCode, e.Message)
}

// GetSecret retrieves a secret by name.
func (c *Client) GetSecret(name, projectID, environment string) (*Secret, error) {
	if environment == "" {
		environment = "prod"
	}

	data, err := c.request("GET", "/api/secrets/by-name/"+url.PathEscape(name), nil, map[string]string{
		"project_id":  projectID,
		"environment": environment,
	})
	if err != nil {
		return nil, err
	}

	var secret Secret
	if err := json.Unmarshal(data, &secret); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &secret, nil
}

// ListSecrets lists secrets in a project.
func (c *Client) ListSecrets(projectID, environment string) ([]Secret, error) {
	params := map[string]string{
		"project_id": projectID,
	}
	if environment != "" {
		params["environment"] = environment
	}

	data, err := c.request("GET", "/api/secrets", nil, params)
	if err != nil {
		return nil, err
	}

	var result struct {
		Data []Secret `json:"data"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return result.Data, nil
}

// CreateSecret creates a new secret.
func (c *Client) CreateSecret(name, value, projectID, environment, note string) (*Secret, error) {
	if environment == "" {
		environment = "prod"
	}

	body := map[string]interface{}{
		"name":       name,
		"value":      value,
		"project_id": projectID,
		"environment": environment,
	}
	if note != "" {
		body["note"] = note
	}

	data, err := c.request("POST", "/api/secrets", body, nil)
	if err != nil {
		return nil, err
	}

	var secret Secret
	if err := json.Unmarshal(data, &secret); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &secret, nil
}

// UpdateSecret updates an existing secret.
func (c *Client) UpdateSecret(secretID string, value, note *string) (*Secret, error) {
	body := map[string]interface{}{}
	if value != nil {
		body["value"] = *value
	}
	if note != nil {
		body["note"] = *note
	}

	data, err := c.request("PUT", "/api/secrets/"+secretID, body, nil)
	if err != nil {
		return nil, err
	}

	var secret Secret
	if err := json.Unmarshal(data, &secret); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &secret, nil
}

// DeleteSecret deletes a secret.
func (c *Client) DeleteSecret(secretID string) error {
	_, err := c.request("DELETE", "/api/secrets/"+secretID, nil, nil)
	return err
}

// ListProjects lists all projects.
func (c *Client) ListProjects() ([]Project, error) {
	data, err := c.request("GET", "/api/projects", nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Data []Project `json:"data"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return result.Data, nil
}

// GetProject gets a project by ID.
func (c *Client) GetProject(projectID string) (*Project, error) {
	data, err := c.request("GET", "/api/projects/"+projectID, nil, nil)
	if err != nil {
		return nil, err
	}

	var project Project
	if err := json.Unmarshal(data, &project); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &project, nil
}

// HealthCheck checks server health.
func (c *Client) HealthCheck(detailed bool) (*HealthStatus, error) {
	params := map[string]string{}
	if detailed {
		params["detailed"] = "true"
	}

	data, err := c.request("GET", "/health", nil, params)
	if err != nil {
		return nil, err
	}

	var status HealthStatus
	if err := json.Unmarshal(data, &status); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &status, nil
}
